app/cdap/components/PipelineDiff/util/diff.ts

```diff
@@ -0,0 +1,138 @@
+}
+
+export function getStageNameFromStageDiffKey(stageKey: string) {
+  return stageKey
+    .split('__')
+    .slice(1)
+    .join('__');
+}
+
+export function getPluginNameFromStageDiffKey(stageKey: string) {
+  return stageKey.split('__')[0];
+}
+
+export function getConnectionDiffKey(
+  connection: IPipelineConnection,
+  getStageDiffKeyFromPluginName: (name: string) => string
+) {
+  const fromName = getStageDiffKeyFromPluginName(connection.from);
+  const toName = getStageDiffKeyFromPluginName(connection.to);
+  return `${fromName}\u0001${toName}`;
+}
+
+export function getStageDiffKeysFromConnectionDiffKey(connectionKey: string) {
+  // returns [fromStageDiffKey, toStageDiffKey]
+  return connectionKey.split('\u0001');
+}
+
+function preprocessPipeline(pipeline: IPipelineConfig) {
+  const stageMap: IStageMap = {};
+  const connectionMap: IConnectionMap = {};
+  const stageNameToStage = {};
+
+  for (const stage of pipeline.stages) {
+    const stageMapKey = `${stage.plugin.name}__${stage.name}`;
+    stageMap[stageMapKey] = stage;
+    stageNameToStage[stage.name] = stage;
+  }
+  const getStageDiffKeyFromPluginName = (name: string) => getStageDiffKey(stageNameToStage[name]);
+
+  for (const connection of pipeline.connections) {
+    const connectionMapKey = getConnectionDiffKey(connection, getStageDiffKeyFromPluginName);
+    connectionMap[connectionMapKey] = connection;
+  }
+  return { stageMap, connectionMap };
+}
+
+function jsonObjectDiff<T1>(obj1: T1, obj2: T1) {
+  return diff(obj1, obj2) as { [name: string]: DeepPartial<T1> };
+}
+
+/**
+ * Given two pipelines, return the difference between them
+ * @param pipeline1 first pipeline
+ * @param pipeline2 second pipeline
+ * @returns edit actions applied to pipeline1 to get pipeline2
+ */
+export function computePipelineDiff(pipeline1: IPipelineConfig, pipeline2: IPipelineConfig) {
+  const { stageMap: stageMap1, connectionMap: connectionMap1 } = preprocessPipeline(pipeline1);
+  const { stageMap: stageMap2, connectionMap: connectionMap2 } = preprocessPipeline(pipeline2);
+
+  const stagesDiffMap = jsonObjectDiff(stageMap1, stageMap2);
+  const connectionsDiffMap = jsonObjectDiff(connectionMap1, connectionMap2);
+
+  const diffMap: IPipelineDiffMap = {
+    stages: {},
+    connections: {},
+  };
+
+  for (const stage in stagesDiffMap) {
+    let stageName = stage;
+    let diffIndicator: DiffIndicator = `~`;
+
+    if (stage.match(/__added$/)) {
+      stageName = stage.replace(/__added$/, '');
+      diffIndicator = '+';
+    } else if (stage.match(/__deleted$/)) {
+      stageName = stage.replace(/__deleted$/, '');
+      diffIndicator = '-';
+    }
+    diffMap.stages[stageName] = {
+      diffIndicator,
+      stage1: stageMap1[stageName],
+      stage2: stageMap2[stageName],
+      diff: stagesDiffMap[stage],
+    };
+  }
+
+  for (const connection in connectionsDiffMap) {
+    let connectionName = connection;
+    let diffIndicator: DiffIndicator = '~';
```

Comment:
Would prefer these diff indicators to be constants rather than strings - I think most of the strings used as separators would benefit from a readability standpoint if they were the same.